File: backend/schemas/template.py
"""
회의록 양식 요청/응답 Pydantic v2 스키마
REQ-TMPL-001: 양식 업로드 응답
REQ-TMPL-003: 양식 목록/상세/삭제 응답
"""

from datetime import datetime

from pydantic import BaseModel, Field


class TemplateUploadResponse(BaseModel):
    """POST /api/v1/templates 응답 (REQ-TMPL-001)"""

    # 양식 고유 ID
    template_id: str = Field(..., description="양식 고유 ID")
    # 양식 이름
    name: str = Field(..., description="양식 이름")
    # 파일 형식 (docx, pdf)
    format: str = Field(..., description="파일 형식 (docx/pdf)")
    # 파싱된 구조
    structure: dict = Field(..., description="양식 구조 (섹션, 필드, 테이블 여부)")
    # 생성 시각
    created_at: str = Field(..., description="업로드 시각 (ISO 8601)")


class TemplateListItem(BaseModel):
    """GET /api/v1/templates 목록 항목 (REQ-TMPL-003)"""

    # 양식 고유 ID
    template_id: str = Field(..., description="양식 고유 ID")
    # 양식 이름
    name: str = Field(..., description="양식 이름")
    # 파일 형식
    format: str = Field(..., description="파일 형식 (docx/pdf)")
    # 생성 시각
    created_at: str = Field(..., description="업로드 시각 (ISO 8601)")


class TemplateDetail(BaseModel):
    """GET /api/v1/templates/{template_id} 상세 응답 (REQ-TMPL-003)"""

    # 양식 고유 ID
    template_id: str = Field(..., description="양식 고유 ID")
    # 양식 이름
    name: str = Field(..., description="양식 이름")
    # 파일 형식
    format: str = Field(..., description="파일 형식 (docx/pdf)")
    # 파싱된 구조
    structure: dict = Field(..., description="양식 구조 (섹션, 필드, 테이블 여부)")
    # 생성 시각
    created_at: str = Field(..., description="업로드 시각 (ISO 8601)")

File: client/lib/models/template.dart
// 양식(Template) 데이터 모델 - SPEC-TMPL-001 REQ-TMPL-005
class Template {
  // 백엔드 API의 template_id 필드
  final String templateId;

  // 파일 원본 이름 (표시용)
  final String name;

  // 파일 형식: 'pdf' 또는 'docx'
  final String format;

  // 양식 구조 정보 (상세 조회 시에만 포함)
  final Map<String, dynamic>? structure;

  // 업로드 일시
  final DateTime createdAt;

  const Template({
    required this.templateId,
    required this.name,
    required this.format,
    this.structure,
    required this.createdAt,
  });

  // 특정 필드만 변경한 복사본 반환
  Template copyWith({
    String? templateId,
    String? name,
    String? format,
    Map<String, dynamic>? structure,
    DateTime? createdAt,
  }) {
    return Template(
      templateId: templateId ?? this.templateId,
      name: name ?? this.name,
      format: format ?? this.format,
      structure: structure ?? this.structure,
      createdAt: createdAt ?? this.createdAt,
    );
  }

  // JSON에서 Template 객체 생성
  factory Template.fromJson(Map<String, dynamic> json) {
    return Template(
      templateId: json['template_id'] as String,
      name: json['name'] as String,
      format: json['format'] as String,
      structure: json['structure'] as Map<String, dynamic>?,
      createdAt: DateTime.parse(json['created_at'] as String),
    );
  }

  // Template 객체를 JSON으로 변환
  Map<String, dynamic> toJson() {
    return {
      'template_id': templateId,
      'name': name,
      'format': format,
      'structure': structure,
      'created_at': createdAt.toIso8601String(),
    };
  }
}
